Create Mod Browser

ThePythonGuy3 opened the pull request:
I know you had the idea of making one, but you said no finally. I decided to try to add it, and this is the result. I added a button to Import Mod. It is missing a bundle, but thats an easy fix. They way it works is simple: 

1. You click that button
2. You scroll down to the mod you want to download
3. Click it. Thats it, and it gets downloaded automatically

TODO:

- Add a search bar
- Add some sort options


This is how the menu looks (It is an old version, that has Author, name, description and stars scrambled, as the newest version refused to get screenshotted):

![image](https://raw.githubusercontent.com/ThePythonGuy3/F/master/unknown.png)

ThePythonGuy3 commented:
Update: I managed to screenshot the newer version, here's how it looks like:

![image](https://raw.githubusercontent.com/ThePythonGuy3/F/master/unknown2.png)

Pietro303HD commented:
Will it need an Internet connection?

ThePythonGuy3 commented:
Of course. It gets the information from github directly.

markozajc commented:
It would probably be a good idea to cache entries for performance.

DeltaNedas commented:
routorio

ThePythonGuy3 commented:
Thanks, very useful @DeltaNedas

TheSlaylord commented:
What happened to mod name?
![IMG_20200731_180911](https://user-images.githubusercontent.com/62336673/89048800-01903700-d359-11ea-9943-ff44d2ef2d91.png)

sk7725 commented:
use labelWrap so the text doesnt go over the box

sk7725 commented:
Also mark this PR for draft till it is ready

sk7725 commented:
Make starring the mods easier and recommended

ThePythonGuy3 commented:
1. i used labelWrap, it didnt work, so i just didnt add it to the PR
2. It is finished. Finally
3. What?

ThePythonGuy3 commented:
> What happened to mod name?
> ![IMG_20200731_180911](https://user-images.githubusercontent.com/62336673/89048800-01903700-d359-11ea-9943-ff44d2ef2d91.png)

It gets the repository link name

sk7725 commented:
Then maybe make the box bigger? Its so ugly that way that I doubt Anuke will add it
Also, something like a button that leads to the github page?
~~something like the Rate Us in other games~~

ThePythonGuy3 commented:
Ok, those ideas are great. Lemme mark this as a draft while i add those

ThePythonGuy3 commented:
Okay, I finished the browser. Some friends and I added bundles for some languages.

Updates:
 - Added a search bar
 - Redid the UI
 - Fixed the Layout
 - Made it more stable
 - Added bundles.
 - Added Github Page Opening button

This is how it looks now:
![image](https://raw.githubusercontent.com/ThePythonGuy3/F/master/unknown%20(1).png)

And this is how the selection dialog looks like:

![image](https://raw.githubusercontent.com/ThePythonGuy3/F/master/unknown%20(2).png)

Credits:
 - @sk7725 Korean Translation
 - @feu-follet French Translation
 - @ThePythonGuy3 & @feu-follet Spanish Translation
 - @ThePythonGuy3 Mod Browser & English Bundle

ThePythonGuy3 commented:
Okay, last update.
 - Added icons to the rest of the buttons.

ThePythonGuy3 commented:
Made this PR compatible with #2305

Anuken commented on core/src/mindustry/ui/dialogs/ModsDialog.java:
```diff
@@ -133,11 +144,121 @@ void setup(){
+                                    tablebrow.clear();
+                                    String searchString = searchtxt.toLowerCase();
+                                    Core.net.httpGet("https://raw.githubusercontent.com/Anuken/MindustryMods/master/mods.json", response -> {
+                                        if (response.getStatus() != HttpStatus.OK) {
```
The formatting in this file is incorrect. You're not following the [style guidelines](https://github.com/Anuken/Mindustry/blob/master/CONTRIBUTING.md#style-guidelines).

ThePythonGuy3 replied:
I dont really know which im violating. If you are referring to tabs instead of 4 spaces i can change it, but i think everything else is good

ThePythonGuy3 replied:
Oh okay, you meant if(cond) {}

Anuken commented on core/src/mindustry/ui/dialogs/ModsDialog.java:
```diff
@@ -133,11 +144,121 @@ void setup(){
+                                    Core.net.httpGet("https://raw.githubusercontent.com/Anuken/MindustryMods/master/mods.json", response -> {
+                                        if (response.getStatus() != HttpStatus.OK) {
+                                            ui.showErrorMessage(Core.bundle.format("connectfail", response.getStatus()));
+                                        } else {
```
This should be wrapped in a `try/catch` with a log, just in case a parsing error occurs.

Anuken commented on core/src/mindustry/ui/dialogs/ModsDialog.java:
```diff
@@ -133,11 +144,121 @@ void setup(){
+                                                    btn.margin(12f);
+                                                    btn.table(con -> {
+                                                        con.left();
+                                                        con.add("[lightgray]Name:[] " + modsbrolist.name + "\n[lightgray]Author:[] " + modsbrolist.author + "\n[accent]\uE809: " + modsbrolist.stars).width(388f).wrap().growX().pad(0f,6f,0f,6f);
```
The star icon doesn't need a colon after it.

Anuken commented on core/src/mindustry/ui/dialogs/ModsDialog.java:
```diff
@@ -133,11 +144,121 @@ void setup(){
+                                                    }).fillY().growX().pad(0f,6f,0f,6f);
+                                                }, Styles.clearPartialt, () -> {
+                                                    BaseDialog modSelected = new BaseDialog((Core.bundle.get("mods.browser.selected"))+ ": "+ modsbrolist.name);
+                                                    modSelected.cont.add(modsbrolist.description).width(mobile ? 400f : 500f).wrap().pad(4f).get().setAlignment(Align.center, Align.center);
```
Don't display the description in the mod list. It should only be visible when clicking the mod.

ThePythonGuy3 replied:
It _is_ only shown when clicked

Anuken replied:
My bad, I didn't see the more recent preview

Anuken commented on core/src/mindustry/ui/dialogs/ModsDialog.java:
```diff
@@ -133,11 +144,121 @@ void setup(){
+                                        } else {
+                                            Json json = new Json();
+                                            Seq<ModListing> listings = json.fromJson(Seq.class, ModListing.class, response.getResultAsString());
+                                            for (ModListing modsbrolist : listings) {
```
Why is this variable called `modsbrolist`? `list` or `mod` would be preferable.

ThePythonGuy3 replied:
Yeah, it was the short of mods browser list, ill change it

Anuken commented on core/src/mindustry/ui/dialogs/ModsDialog.java:
```diff
@@ -133,11 +144,121 @@ void setup(){
+                            dialog2.addCloseButton();
+                            dialog2.show();
+                            //Keep this until mods work
+                            ui.showErrorMessage("This feature doesnt work completely due to mods not working in v6. Wait for this to be fixed. If you decide to use it anyways, mods will not let you enter the game.");
```
This should not be necessary. Incompatible mods (e.g. with min-game-version < 105) should not be installable at all.

ThePythonGuy3 replied:
Most mods dont work at all in v6, but work in v104.6, so u added the warning. If you think it's unnecessary ill delete it

Anuken replied:
I'm saying the mod browser should display incompatible mods instead of a startup dialog. You may need to wait until I make incompatibility more explicit for mods.

ThePythonGuy3 replied:
Okay, then i delete the message and you'll do whatever you need to do

ThePythonGuy3 commented:
@Anuken Now, if i didn't miss anything, this is all you asked for. If something is missing or needs to be deleted let me know.

Remint32 commented:
@ThePythonGuy3 In **6.0**, we have icons for mods. You may display the `icon.png` file from the root of the github modification repository, if there is such a file. Otherwise display the placeholder.

ThePythonGuy3 commented:
@Remint32 I cant do that really without downloading the image or the mod itself, and thats not the point. i just get the info from Anuken/Mindustry-Mods, so i get limited info (name, repo, stars, etc..) but no icons. Thats a file inside the repo. Although i could download that image and store it somewhere in the savefiles, thats not precisely cheap on mobile data, to download the image of each mod. (Nor i want to download each mod, since then the browser would have no point to exist)

Remint32 commented:
@ThePythonGuy3 I will think about these points and write

ThePythonGuy3 commented:
Thank you

Anuken commented on core/src/mindustry/ui/dialogs/ModsDialog.java:
```diff
@@ -133,11 +144,118 @@ void setup(){
+                                        }else{
+                                            Json json = new Json();
+                                            Seq<ModListing> listings = json.fromJson(Seq.class, ModListing.class, response.getResultAsString());
+                                            if(ModListing mod : listings){
```
This, uh, doesn't compile.

ThePythonGuy3 replied:
What? OhnoOhnoOhnoOhno. What does intellij say?

ThePythonGuy3 replied:
Its very strange since it compiles in my case :/

ThePythonGuy3 replied:
Whats does intellij say?

Anuken replied:
There's no way `if(ModListing mod : listings){` compiles, that statement doesn't make any sense. You're confusing an if with a `for`.

ThePythonGuy3 replied:
ohno, somehow i managed to change the fors with ifs, but in the version on my PC isnt, very strange

ThePythonGuy3 commented:
Fixed

Anuken commented:
I'll take another look at this when the mod updater is back online.

ThePythonGuy3 commented:
okay, its a nice idea

Shawak commented on core/src/mindustry/ui/dialogs/ModsDialog.java:
```diff
@@ -133,11 +144,118 @@ void setup(){
+                                                        modSelected.hide();
+                                                        installMod[0] = () -> {
+                                                            ui.loadfrag.show();
+                                                            Core.net.httpGet("http://api.github.com/repos/" + mod.repo + "/zipball/master", loc -> {
```
https

ThePythonGuy3 replied:
h

ThePythonGuy3 replied:
anuke doesnt do that

LottieVixen replied:
they actually do in most places 
![image](https://user-images.githubusercontent.com/495512/101247164-93e18680-370f-11eb-8f0e-ab8e952050bc.png)

Anuken commented:
This has been merged into a local branch, and does not need to stay open anymore. I've cleaned up much of the code, but there's still quite a lot of work needed to get this into the game.
